File: Data Structures & Algorithms/longest-consecutive-sequence/submission-1.cpp
class Solution {
public:
    unordered_map<int, int> best_sequence;
    unordered_set<int> present;

    int find_best_sequence(int x) {
        if (present.count(x) == 0) {
            return 0;
        }

        if (best_sequence.find(x) != best_sequence.end()) {
            return best_sequence[x];
        } 

        return best_sequence[x] = find_best_sequence(x+1)+1;
    }

    int longestConsecutive(vector<int>& nums) {
        for (auto num : nums) {
            present.insert(num);
        }

        int best = 0;

        for (auto num : nums) {
            best = max(best, find_best_sequence(num));
        }

        return best;
    }
};


/*

    best_sequence[i] = longest current sequence that can be formed starting with i

    



    [2,20,4,10,3,4,5]

    best_sequence[5] = 1;

    best_sequence[4] = 2;

    best_sequence[i] = best_sequence[i+1] + 1;

*/

File: Data Structures & Algorithms/top-k-elements-in-list/submission-1.py
class Solution:
    def topKFrequent(self, nums: List[int], k: int) -> List[int]:
        freq = [0] * 2010

        for num in nums:
            freq[num + 1000] += 1

        l = []
        for i in range(2010):
            if(freq[i] > 0):
                l.append((freq[i], i))

        out = []

        l = sorted(l)[::-1]

        for i in range(k):
            out.append(l[i][1] - 1000)

        return out
